Add ip location cache

downdawn opened the pull request.

downdawn commented:
PLR.

wu-clan commented on backend/app/core/conf.py:
```diff
@@ -144,6 +144,10 @@ def validator_api_url(cls, values):
     OPERA_LOG_ENCRYPT_INCLUDE: list[str] = ['password', 'old_password', 'new_password', 'confirm_password']
 
+    # ip location
+    IP_LOCATION_REDIS_PREFIX: str = 'ip_location'
```
We should use the `fba` prefix to ensure a unified style.

wu-clan commented on backend/app/utils/request_parse.py:
```diff
@@ -69,18 +70,24 @@ def get_location_offline(ip: str) -> list[str] | None:
+                country = location_info[0] if location_info[0] != '0' else None
+                region = location_info[2] if location_info[2] != '0' else None
+                city = location_info[3] if location_info[3] != '0' else None
+    await redis_client.set(f'{settings.IP_LOCATION_REDIS_PREFIX}:{ip}', f'{country} {region} {city}',
```
It should be put in else.

wu-clan commented on backend/app/utils/request_parse.py:
```diff
@@ -69,18 +70,24 @@ def get_location_offline(ip: str) -> list[str] | None:
+    if location:
+        country, region, city = location.split(' ')
+    else:
+        if settings.LOCATION_PARSE == 'online':
```
Do you want to try match-case?

downdawn replied:
Match-case doesn't seem to be appropriate here?

wu-clan replied:
yes.

wu-clan reviewed (changes requested):
please update.

downdawn commented:
```
@sync_to_async
def get_location_offline(ip: str) -> dict | None:
    """
    离线获取 ip 地址属地，无法保证准确率，100%可用

    :param ip:
    :return:
    """
    try:
        cb = XdbSearcher.loadContentFromFile(dbfile=IP2REGION_XDB)
        searcher = XdbSearcher(contentBuff=cb)
        data = searcher.search(ip)
        searcher.close()
        data = data.split('|')
        return {'country': data[0], 'regionName': data[2], 'city': data[3]}
    except Exception as e:
        log.error(f'离线获取 ip 地址属地失败，错误信息：{e}')
        return None


async def parse_ip_info(request: Request) -> tuple[str, str, str, str]:
    country, region, city = None, None, None
    ip = await get_request_ip(request)
    location = await redis_client.get(f'{settings.IP_LOCATION_REDIS_PREFIX}:{ip}')
    if location:
        country, region, city = location.split(' ')
        return ip, country, region, city
    if settings.LOCATION_PARSE == 'online':
        location_info = await get_location_online(ip, request.headers.get('User-Agent'))
    elif settings.LOCATION_PARSE == 'offline':
        location_info = await get_location_offline(ip)
    else:
        location_info = None
    if location_info:
        country = location_info.get('country')
        region = location_info.get('regionName')
        city = location_info.get('city')
    await redis_client.set(f'{settings.IP_LOCATION_REDIS_PREFIX}:{ip}', f'{country} {region} {city}',
                           ex=settings.IP_LOCATION_EXPIRE_SECONDS)
    return ip, country, region, city
```

downdawn commented:
Update.

wu-clan reviewed (approved):
LGTM.
